### ISSUE TITLE
new lint: `suspicious_matches`

### ISSUE THREAD

**Issue by y21** [rust-lang/rust-clippy#10640]
### What it does

Looks for `matches!` invocations that look like they have the "argument" order swapped. One such case would be having a binding pattern without a match guard. It always matches any value and introduces a binding that you cannot refer to.

```rs
#![allow(unused)]
enum Foo {
  Bar,
  Baz
}

fn is_bar(v: Foo) -> bool {
  matches!(Foo::Bar, v)
}
```

The intent was most likely to match on the value stored in `v` and check if it matches `Foo::Bar`, but the order of arguments is wrong. `Foo::Bar` happens to also be a valid expression and `v` happens to be a valid pattern.
The correct way to write this would be:
```rs
fn is_bar(v: Foo) -> bool {
  matches!(v, Foo::Bar)
}
```

### Lint Name

suspicious_matches

### Category

suspicious

### Advantage

Original code did not actually check anything. This lint makes the user aware of it and suggests swapping the value and pattern

### Drawbacks

None that I can think of

### Example

(see description)

**Comment by y21** [comment]
Woops, this looks like a duplicate of #7351
